cpp/src/experimental/louvain.cuh

```diff
@@ -1205,350 +1207,49 @@ class Louvain {
-      // partitions will actually have data.
-      //
-      rmm::device_vector<int> partition_v(new_src_v.size());
-
-      thrust::transform(
-        rmm::exec_policy(stream_)->on(stream_),
-        thrust::make_zip_iterator(thrust::make_tuple(new_src_v.begin(), new_dst_v.begin())),
-        thrust::make_zip_iterator(thrust::make_tuple(new_src_v.end(), new_dst_v.end())),
-        partition_v.begin(),
-        [d_edge_device_view = compute_partition_.edge_device_view()] __device__(
-          thrust::tuple<vertex_t, vertex_t> tuple) {
-          return d_edge_device_view(thrust::get<0>(tuple), thrust::get<1>(tuple));
-        });
-
-      new_src_v = variable_shuffle<graph_view_t::is_multi_gpu, vertex_t>(
-        handle_, partition_v.size(), new_src_v.begin(), partition_v.begin());
-
-      new_dst_v = variable_shuffle<graph_view_t::is_multi_gpu, vertex_t>(
-        handle_, partition_v.size(), new_dst_v.begin(), partition_v.begin());
-
-      new_weight_v = variable_shuffle<graph_view_t::is_multi_gpu, weight_t>(
-        handle_, partition_v.size(), new_weight_v.begin(), partition_v.begin());
-
-      //
-      //  Now everything is on the correct node, again combine like edges
-      //
-      std::tie(new_src_v, new_dst_v, new_weight_v) =
-        combine_local_edges(new_src_v, new_dst_v, new_weight_v);
-    }
-
-    //
-    //  Now I have a COO of the new graph, distributed according to the
-    //  original clustering (eventually this likely fits on one GPU and
-    //  everything else is empty).
-    //
-    current_graph_ =
-      detail::create_graph<vertex_t,
-                           edge_t,
-                           weight_t,
-                           graph_t::is_adj_matrix_transposed,
-                           graph_t::is_multi_gpu>(handle_,
-                                                  new_src_v,
-                                                  new_dst_v,
-                                                  new_weight_v,
-                                                  num_clusters,
-                                                  experimental::graph_properties_t{true, true},
-                                                  current_graph_view_);
+    std::tie(current_graph_, numbering_map) =
+      coarsen_graph(handle_, current_graph_view_, dendrogram_->current_level_begin());
 
     current_graph_view_ = current_graph_->view();
 
-    src_indices_v_.resize(new_src_v.size());
-
     local_num_vertices_ = current_graph_view_.get_number_of_local_vertices();
     local_num_rows_     = current_graph_view_.get_number_of_local_adj_matrix_partition_rows();
     local_num_cols_     = current_graph_view_.get_number_of_local_adj_matrix_partition_cols();
-    local_num_edges_    = new_src_v.size();
+    base_vertex_id_     = current_graph_view_.get_local_vertex_first();
+
+    local_num_edges_ = edge_t{0};
+    for (size_t i = 0; i < current_graph_view_.get_number_of_local_adj_matrix_partitions(); ++i) {
```

Comment:
Consider replacing custom loop by algorithm (in this case, probably `std::accumulate()`).